runtime/jcl/common/java_lang_Class.cpp

```diff
@@ -1827,6 +1970,20 @@ Java_java_lang_Class_getNestHostImpl(JNIEnv *env, jobject recv)
 
 	J9Class *clazz = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, J9_JNI_UNWRAP_REFERENCE(recv));
 	J9Class *nestHost = clazz->nestHost;
+
+	/*
+	 * Nest host field is set post-verification, during an access check or during
+	 * a call to relevant java.lang.Class API method via the loadAndCheckNestHost
+	 * function. If loadAndCheckNestHost fails to return a valid nest host, then
+	 * getNestHostImpl returns this.
+	 */
+	if (NULL == nestHost) {
+		nestHost = loadAndCheckNestHost(currentThread, clazz, FALSE);
```

Comment:
does OpenJDK load classes when getNestHost\getNestMembers is called?
If it does, what happens if there is an error verifying nest membership?

Reply:
I ran `java -verbose:class` against an older OpenJDK nestmates build and against this for the valid test case:
```
public class nestmates_api_tests_delayed_loading {
	public class NestMember {}
	public class NestHost {}
	
	public static void main (String[] args) {
		System.out.println("Nest: " + Arrays.toString(new nestmates_api_tests_delayed_loading().new NestHost().getClass().getClasses()));
	}
}
```

OpenJDK seems to load classes; output produced at the end was:
```
[0.237s][info][class,load] nestmates_api_tests_delayed_loading source: file:/team/tdmccormick/projects/20170509-NestMates/
[0.237s][info][class,load] java.lang.NamedPackage source: jrt:/java.base
[0.238s][info][class,load] java.lang.PublicMethods$MethodList source: jrt:/java.base
[0.238s][info][class,load] java.lang.PublicMethods$Key source: jrt:/java.base
[0.238s][info][class,load] java.lang.Void source: jrt:/java.base
[0.240s][info][class,load] nestmates_api_tests_delayed_loading$NestHost source: file:/team/tdmccormick/projects/20170509-NestMates/
[0.242s][info][class,load] nestmates_api_tests_delayed_loading$NestMember source: file:/team/tdmccormick/projects/20170509-NestMates/
Nest: [class nestmates_api_tests_delayed_loading$NestHost, class nestmates_api_tests_delayed_loading$NestMember, class nestmates_api_tests_delayed_loading]
[0.243s][info][class,load] java.lang.Shutdown source: jrt:/java.base
[0.244s][info][class,load] java.lang.Shutdown$Lock source: jrt:/java.base
```

This was similar; output produced at the end was:
```
class load: nestmates_api_tests_delayed_loading from: file:/team/tdmccormick/projects/20170509-NestMates/
class load: jdk/internal/reflect/UnsafeFieldAccessorFactory from: jrt:/java.base
class load: jdk/internal/reflect/FieldAccessor from: jrt:/java.base
class load: jdk/internal/reflect/FieldAccessorImpl from: jrt:/java.base
class load: jdk/internal/reflect/UnsafeFieldAccessorImpl from: jrt:/java.base
class load: jdk/internal/reflect/UnsafeObjectFieldAccessorImpl from: jrt:/java.base
class load: java/lang/Class$ReflectCache from: nestmates_api_patch_3.jar
class load: nestmates_api_tests_delayed_loading$NestHost from: file:/team/tdmccormick/projects/20170509-NestMates/
class load: nestmates_api_tests_delayed_loading$NestMember from: file:/team/tdmccormick/projects/20170509-NestMates/
Nest: [class nestmates_api_tests_delayed_loading$NestHost, class nestmates_api_tests_delayed_loading$NestMember, class nestmates_api_tests_delayed_loading]
class load: java/util/IdentityHashMap$KeySet from: jrt:/java.base
class load: java/util/IdentityHashMap$IdentityHashMapIterator from: jrt:/java.base
class load: java/util/IdentityHashMap$KeyIterator from: jrt:/java.base
class load: com/ibm/tools/attach/target/AttachHandler$1 from: jrt:/java.base
class load: java/util/Vector$Itr from: jrt:/java.base
class load: java/lang/Thread$State from: jrt:/java.base
```
(the `nestmates_api_patch_3.jar` is my patch for testing)

Reply:
**QUESTION: Should I change our error message so we don't exactly match OpenJDK?**

Running the same test case but invalid nestmates attributes (one of the listed nest members does not claim the nest host) raises an error for both OpenJDK and this.

OpenJDK:
```
/team/tdmccormick/jvms/Nestmates_Build_15.01.2018/jdk/bin/java  nestmates_api_tests_delayed_loading
Exception in thread "main" java.lang.IncompatibleClassChangeError: Nest member nestmates_api_tests_delayed_loading in nestmates_api_tests_delayed_loading$NestHost declares a different nest host of nestmates_api_tests_delayed_loading
        at java.base/java.lang.Class.getNestMembers0(Native Method)
        at java.base/java.lang.Class.getNestMembers(Class.java:3922)
        at nestmates_api_tests_delayed_loading.main(Unknown Source)
```

This:
```

Exception in thread "main" java.lang.IncompatibleClassChangeError: Nest member nestmates_api_tests_delayed_loading in nestmates_api_tests_delayed_loading$NestHost declares a different nest host of nestmates_api_tests_delayed_loading
        at java.lang.Class.getNestMembersImpl(java.base@9-internal/Native Method)
        at java.lang.Class.getNestMembers(java.base@9-internal/Class.java:4126)
        at nestmates_api_tests_delayed_loading.main(Unknown Source)
```

Reply:
I get two copies of the same IncompatibleClassChangeError (ICCE) from calling `getNestMembers` and none from `getNestHost`. This behaviour is expected. `getNestMembers` fails because the nest host iterates over its list of nest members and checks that they reciprocally verify eachother; it raises an ICCE when it finds that a nest member does not verify the nest host. `getNestHost` passes because the nest member class considers itself to be its own nest host and thus does no further checking.

--

The test case ran:
```
for (int x = 0; x < 2; x++) {
	try {
		System.out.println("Nest host: " + nestMemberClass.getNestHost());
	} catch (IncompatibleClassChangeError e) {
		System.out.println(e + " caused by " + e.getCause());
	}
}		
for (int x = 0; x < 2; x++) {
	try {
		System.out.println("Nest: " + Arrays.toString(nestHostClass.getNestMembers()));
	} catch (IncompatibleClassChangeError e) {
		System.out.println(e + " caused by " + e.getCause());
	}
}
```
and it produced
```
Nest host: class pkg.NestMember
Nest host: class pkg.NestMember
java.lang.IncompatibleClassChangeError: Nest member pkg/NestMember in pkg/NestMember$NestHost declares a different nest host of pkg/NestMember caused by null
java.lang.IncompatibleClassChangeError: Nest member pkg/NestMember in pkg/NestMember$NestHost declares a different nest host of pkg/NestMember caused by null
```

Reply:
Triggering an IncompatibleClassChangeError from `getNestHost` (ie. the nest member class claims a nest host which does not reciprocally verify it) is not possible. If any verification error occurs, then the class itself is returned. 

--

The test case ran:
```
for (int x = 0; x < 2; x++) {
	try {
		System.out.println("Nest host: " + nestMemberClass.getNestHost());
	} catch (IncompatibleClassChangeError e) {
		System.out.println(e + " caused by " + e.getCause());
	}
}
```
and it produced
```
Nest host: class pkg.NestMember
Nest host: class pkg.NestMember
```
despite claiming a different class as its nest host. This lines up with the current spec draft: 
>  Returns:
    the nest host of this class, or this if we cannot obtain a valid nest host 

(http://cr.openjdk.java.net/%7Edholmes/8010319/specs/java.lang/java/lang/Class.html)

and also matches the behaviour of a nestmates openjdk build.